### PR TITLE
Add nodejs reqs

### PR DESCRIPTION
**Pull request by fredex42**
- adding requirements for nodejs to packer build
- datastore implementation for nodejs
- test suite for nodejs datastore implementation
- adding circleci tests for datastore implementation

The datastore implementation makes extensive use of Promises to make it non-blocking, each get, substitute, etc. returns a Promise object which can then be handled using the standard .done(), .then() etc. functions

### CODE REVIEW

**Comment by fredex42** [comment]
@clloyd @currysoup can you review pls? thanks

File: CDS/js/Datastore.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            });
+        });
+    },
+    close: function() {
+        db.close();
+    },
+    set: set,
+    setMulti: setMulti,
+    get: get,
+    substituteString: function(conn,str){
```
Comment (clloyd) [inline CDS/js/Datastore.js]: I've put together an alternative substitute string function moving as much as possible into sync replacements and avoiding unneeded promise creation.

```js
function(conn,str){
        return new Promise(function(fulfill,reject) {
            
            function replaceAllOccurances(string, matchText, replacementValue) {
                //This replaces all occurances of matchText (by splitting) then replaces then with the value (by joining)
                const sanitisedReplacement = replacementValue ? replacementValue : "undefined";
                return string.split(matchText).join(sanitisedReplacement);
            }

            function performStaticSubstitutions(str) {
                const d = new Date();
                const static_subs = {
                    '{route-name}': process.env.cf_route_name,
                    '{hostname}': process.env.HOSTNAME,
                    '{ostype}': process.env.OSTYPE,
                    '{year}': d.getFullYear(),
                    '{month}': d.getMonth()+1,
                    '{day}': d.getDay(),
                    '{hour}': d.getHours(),
                    '{min}': d.getMinutes(),
                    '{sec}': d.getSeconds()
                };

                return Object.keys(static_subs).reduce((currentString, staticSubKey) => {
                    return replaceAllOccurances(currentString, staticSubKey, static_subs[staticSubKey])
                }, str)
            }


            function performConfigSubstitutions(str) {
                return Object.keys(conn.staticDefs).reduce((currentString, currentDef) => {
                    const matchText = "{config:" + currentDef + "}"
                    return replaceAllOccurances(currentString, matchText, conn.staticDefs[currentDef] )
                }, str);
            }

            str = performStaticSubstitutions(str);
            str = performConfigSubstitutions(str);


            /* Perform the Async Datastore Replacements */
            var promiseList = [];
            var param_matcher = /\{(\w+):([^}]+)\}/g;

            while(matches = param_matcher.exec(str)){
                var matchtext = matches[0];
                var type=matches[1];
                var key=matches[2];

                if(type !== 'config') {
                    promiseList.push(get(conn, type, key, function (result, type, key, matchtext) {
                        return {'find': matchtext, 'replace': result}
                    }, matchtext));
                } else {
                    promiseList.push(new Promise((fulfill, reject) => {
                        fulfill({'find': matchtext, 'replace': "(value not found)"});
                    }));
                }
            }

            Promise.all(promiseList).done(function(replacementsList){                
                const returnString = replacementsList.reduce((currentString, replacementObj) => {
                    return replaceAllOccurances(currentString, replacementObj.find, replacementObj.replace);
                }, str)
                fulfill(returnString);
            },function(err){
                reject(err);
            });
        });
    }
```

Comment (clloyd) [inline CDS/js/Datastore.js]: @currysoup Any input into making the last little bit more functional?

File: packer/cdsbase/cloud-init-build.sh [hunk 2]
```diff
@@ -38,6 +38,13 @@ apt-get -y install libdbd-mysql-perl
 ./cpanm --force Amazon::SQS::Simple
 ./cpanm Data::UUID URL::Encode Net::FTP::Throttle Digest::SHA1 File::Touch Search::Elasticsearch
 
+echo ------------------------------------------
+echo Kickstarter: Installing JS prerequisites
+echo ------------------------------------------
+npm install sqlite3
```
Comment (clloyd) [inline packer/cdsbase/cloud-init-build.sh]: this is in the `package.json` does it need to be installed separately?

File: circle.yml [hunk 3]
```diff
@@ -11,10 +11,12 @@ checkout:
     - chmod a+x ./circle_setup_rvm.sh
     - sudo -E ./circle_setup_rvm.sh
     - sudo -i /home/ubuntu/content_delivery_system/CDS/install.sh -y
+    - cd CDS/js; npm install
```
Comment (clloyd) [inline circle.yml]: Circle provide a slightly cleaner way of doing this so we don't need the cd https://github.com/guardian/flexible-content/blob/master/circle.yml#L60

File: CDS/js/package.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+  "dependencies": {
+    "mocha": "3.1.2",
+    "sqlite3": "3.1.6",
+    "promise": "7.1.1",
+    "chai-as-promised": "6.0.0",
+    "mkpath": "1.0.0"
+  },
+  "scripts": {
+    "test": "node node_modules/mocha/bin/mocha"
```
Comment (clloyd) [inline CDS/js/package.json]: This can also just be `"test": "mocha"`

**Comment by fredex42** [comment]
@clloyd thanks :)

**Pull request merged**